modules/core/src/main/java/org/apache/ignite/internal/processors/rollingupgrade/RollingUpgradeProcessor.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.ignite.internal.processors.metastorage.ReadableDistributedMetaStorage;
+import org.apache.ignite.internal.util.lang.IgnitePair;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.lang.IgniteProductVersion;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.IgniteNodeAttributes.ATTR_BUILD_VER;
+
+/** Rolling upgrade processor. Manages current and target versions of cluster. */
+public class RollingUpgradeProcessor extends GridProcessorAdapter {
+    /** Key for the distributed property that holds current and target versions. */
+    private static final String ROLL_UP_VERSIONS = "rolling.upgrade.versions";
+
+    /** Metastorage with the write access. */
+    @Nullable private volatile DistributedMetaStorage metastorage;
+
+    /** Distributed property that holds current and target version. */
+    private final AtomicReference<IgnitePair<IgniteProductVersion>> verPairHolder = new AtomicReference<>();
+
+    /**
+     * @param ctx Context.
+     */
+    public RollingUpgradeProcessor(GridKernalContext ctx) {
+        super(ctx);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void start() throws IgniteCheckedException {
+        ctx.internalSubscriptionProcessor().registerDistributedMetastorageListener(new DistributedMetastorageLifecycleListener() {
+            @Override public void onReadyForRead(ReadableDistributedMetaStorage metastorage) {
+                try {
+                    IgnitePair<IgniteProductVersion> pair = metastorage.read(ROLL_UP_VERSIONS);
+
+                    if (verPairHolder.compareAndSet(null, pair) && log.isDebugEnabled())
+                        log.debug("Read current and target versions from metastore: " + pair);
+                }
+                catch (IgniteCheckedException e) {
+                    if (log.isDebugEnabled())
+                        log.debug("Could not read current and target versions: " + e.getMessage());
+                }
+
+                metastorage.listen(ROLL_UP_VERSIONS::equals, (key, oldVal, newVal) ->
+                    verPairHolder.compareAndSet((IgnitePair<IgniteProductVersion>)oldVal, (IgnitePair<IgniteProductVersion>)newVal)
+                );
+            }
+
+            @Override public void onReadyForWrite(DistributedMetaStorage metastorage) {
+                RollingUpgradeProcessor.this.metastorage = metastorage;
+            }
+        });
+    }
+
+    /**
+     * @param current Current version.
+     * @param target Target version.
+     */
+    public void versions(@NotNull IgniteProductVersion current, @NotNull IgniteProductVersion target) throws IgniteCheckedException {
+        A.notNull(metastorage, "Metastorage not ready. Node not started?");
+
+        if (checkVersions(current, target))
+            metastorage.write(ROLL_UP_VERSIONS, F.pair(current, target));
```

Comment:
Should check whether we already is in RU process with other version?

2.18.0 -> 2.18.1
2.18.0 -> 2.19.0